From d73c1c7b834e6e0c91e9028550a1b4db4c793490 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:03:12 -0400
Subject: [usb] wusbcore: return -ENOENT for unlinked URBs

Message-id: <1410887463-20674-121-git-send-email-dzickus@redhat.com>
Patchwork-id: 95190
O-Subject: [RHEL7 PATCH 120/591] usb: wusbcore: return -ENOENT for unlinked URBs.
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit e05a1fd9468bc99bf67bd81601d46d84d93c21c8
Author: Thomas Pugliese <thomas.pugliese@gmail.com>
Date:   Mon Nov 25 16:17:18 2013 -0600

    usb: wusbcore: return -ENOENT for unlinked URBs.

    Return -ENOENT for unlinked URBs.

    Signed-off-by: Thomas Pugliese <thomas.pugliese@gmail.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/wusbcore/wa-xfer.c b/drivers/usb/wusbcore/wa-xfer.c
index 3220c62..a88b8c6 100644
--- a/drivers/usb/wusbcore/wa-xfer.c
+++ b/drivers/usb/wusbcore/wa-xfer.c
@@ -1848,6 +1848,7 @@ int wa_urb_dequeue(struct wahc *wa, struct urb *urb, int status)
   pr_debug("%s: xfer %p id 0x%08X has no RPIPE.  %s",
    __func__, xfer, wa_xfer_id(xfer),
    "Probably already aborted.\n" );
+  result = -ENOENT;
   goto out_unlock;
  }
  /* Check the delayed list -> if there, release and complete */
@@ -1878,6 +1879,7 @@ int wa_urb_dequeue(struct wahc *wa, struct urb *urb, int status)
     * segments will be completed in the DTI interrupt.
     */
    seg->status = WA_SEG_ABORTED;
+   seg->result = -ENOENT;
    spin_lock_irqsave(&rpipe->seg_lock, flags2);
    list_del(&seg->list_node);
    xfer->segs_done++;
@@ -1917,12 +1919,12 @@ int wa_urb_dequeue(struct wahc *wa, struct urb *urb, int status)
   wa_xfer_completion(xfer);
  if (rpipe_ready)
   wa_xfer_delayed_run(rpipe);
- return 0;
+ return result;
 
 out_unlock:
  spin_unlock_irqrestore(&xfer->lock, flags);
 out:
- return 0;
+ return result;
 
 dequeue_delayed:
  list_del_init(&xfer->list_node);
@@ -1958,7 +1960,7 @@ static int wa_xfer_status_to_errno(u8 status)
   [WA_XFER_STATUS_NOT_FOUND] =  0,
   [WA_XFER_STATUS_INSUFFICIENT_RESOURCE] = -ENOMEM,
   [WA_XFER_STATUS_TRANSACTION_ERROR] =  -EILSEQ,
-  [WA_XFER_STATUS_ABORTED] =   -EINTR,
+  [WA_XFER_STATUS_ABORTED] =  -ENOENT,
   [WA_XFER_STATUS_RPIPE_NOT_READY] =  EINVAL,
   [WA_XFER_INVALID_FORMAT] =   EINVAL,
   [WA_XFER_UNEXPECTED_SEGMENT_NUMBER] =  EINVAL,
-- 
1.7.1